contracts/Comptroller.sol

```diff
@@ -1150,151 +1084,196 @@ contract Comptroller is ComptrollerV4Storage, ComptrollerInterface, ComptrollerE
+     * @param cToken The market whose index to update
+     * @param marketState The market state whose index to update
+     * @param vestingState The market vesting state whose index to update
+     * @param isSupply True if this implements the supply update, false if the borrow update
+     * @param marketBorrowIndex The given market's borrow index
+     */
+    function updateCompMarketIndex(address cToken, CompMarketState storage marketState, CompMarketState storage vestingState, bool isSupply, Exp memory marketBorrowIndex) internal {
+        uint compSpeed = compSpeeds[cToken];
         uint blockNumber = getBlockNumber();
-        uint deltaBlocks = sub_(blockNumber, uint(borrowState.block));
-        if (deltaBlocks > 0 && borrowSpeed > 0) {
-            uint borrowAmount = div_(CToken(cToken).totalBorrows(), marketBorrowIndex);
-            uint compAccrued = mul_(deltaBlocks, borrowSpeed);
-            Double memory ratio = borrowAmount > 0 ? fraction(compAccrued, borrowAmount) : Double({mantissa: 0});
-            Double memory index = add_(Double({mantissa: borrowState.index}), ratio);
-            compBorrowState[cToken] = CompMarketState({
-                index: safe224(index.mantissa, "new index exceeds 224 bits"),
-                block: safe32(blockNumber, "block number exceeds 32 bits")
-            });
+        uint deltaBlocks = sub_(blockNumber, uint(marketState.block));
+        if (deltaBlocks > 0 && compSpeed > 0) {
+            uint marketSize;
+            if (isSupply) {
+                marketSize = CToken(cToken).totalSupply();
+            } else {
+                marketSize = div_(CToken(cToken).totalBorrows(), marketBorrowIndex);
+            }
+            uint compAccrued;
+            Double memory ratio;
+            Double memory index;
+
+            if (lastVestingBlock > vestingState.block) {
+                uint deltaVestingBlocks = sub_(lastVestingBlock, uint(marketState.block));
+                compAccrued = mul_(deltaVestingBlocks, compSpeed);
+                ratio = marketSize > 0 ? fraction(compAccrued, marketSize) : Double({mantissa: 0});
+                // important reference to marketState index below as vesting index is not kept "up to date"
+                index = add_(Double({mantissa: marketState.index}), ratio);
+                vestingState.index = safe224(index.mantissa, "new index exceeds 224 bits");
+                vestingState.block = safe32(lastVestingBlock, "block number exceeds 32 bits");
+            }
+
+            compAccrued = mul_(deltaBlocks, compSpeed);
+            ratio = marketSize > 0 ? fraction(compAccrued, marketSize) : Double({mantissa: 0});
+            index = add_(Double({mantissa: marketState.index}), ratio);
+            marketState.index = safe224(index.mantissa, "new index exceeds 224 bits");
+            marketState.block = safe32(blockNumber, "block number exceeds 32 bits");
```

Comment:
On line 1134 an else structure is used to compute the alternative branch variables, but not here. Why is that? That is, in these highlighted lines, does it not re-compute the variables again even if the conditional on 1141 is satisfied? I might understand variable assignments incorrectly in Solidity, but to me it seems that compAccrued, ratio, index is always mandated by this structure therefore the if structure on line 1141 has no effect, as the variables are reassigned here.

Reply:
Hi @jhvst, thanks for flagging this. You're absolutely right, there is some variable reuse here, but it's done for gas reasons and because Solidity puts hard limits on the number of local variables.

The `compAccrued` at line 1143 indeed will get overridden on line 1151, however, it is used to calculate the `vestingState` whereas in line 1151 it is used to calculate the `marketState`.

Agree that the need to reuse variables in Solidity is unfortunate and bad style in general :).